src/plugin_manager/basic_plugin_manager.cpp

```diff
@@ -0,0 +1,157 @@
+/*
+ * Punica - LwM2M server with REST API
+ * Copyright (C) 2019 8devices
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <dlfcn.h>
+#include <iostream>
+
+#include "basic_core.hpp"
+#include "basic_plugin_manager.hpp"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "basic_plugin_manager.h"
+
+basic_plugin_manager_t *basic_plugin_manager_new(basic_punica_core_t *c_core)
+{
+    punica::Core::ptr core(reinterpret_cast<BasicCore *>(c_core));
+
+    return reinterpret_cast<basic_plugin_manager_t *>(new BasicPluginManager(core));
+}
+
+void basic_plugin_manager_delete(basic_plugin_manager_t *c_manager)
+{
+    delete reinterpret_cast<BasicPluginManager *>(c_manager);
+}
+
+int basic_plugin_manager_load_plugin(basic_plugin_manager_t *c_manager,
+                                     const char *c_path,
+                                     const char *c_name)
+{
+    BasicPluginManager *manager = reinterpret_cast<BasicPluginManager *>(c_manager);
+    const std::string path(c_path);
+    const std::string name(c_name);
+
+    return static_cast<int>(manager->loadPlugin(path, name));
+}
+
+int basic_plugin_manager_unload_plugin(basic_plugin_manager_t *c_manager,
+                                       const char *c_name)
+{
+    BasicPluginManager *manager = reinterpret_cast<BasicPluginManager *>(c_manager);
+    const std::string name(c_name);
+
+    return static_cast<int>(manager->unloadPlugin(name));
+}
+
+#ifdef __cplusplus
+} // extern "C"
+#endif
+
+BasicPluginManager::BasicPluginManager(punica::Core::ptr core):
+    mCore(core),
+    mPlugins()
+{ }
+
+BasicPluginManager::~BasicPluginManager()
+{
+    std::map<std::string, std::pair<punica::plugin::Plugin *, punica::plugin::PluginApi *> >::iterator
+    pluginsIterator;
+
+    for (pluginsIterator = mPlugins.begin(); pluginsIterator != mPlugins.end(); ++pluginsIterator)
+    {
+        unloadPlugin(pluginsIterator->first);
+    }
+}
+
+bool BasicPluginManager::loadPlugin(std::string path, std::string name)
+{
+    if (mPlugins.find(name) != mPlugins.end())
+    {
+        if (unloadPlugin(name) != true)
+        {
+            return false;
+        }
+    }
+
+    void *dll = dlopen(path.c_str(), RTLD_NOW | RTLD_LOCAL);
```

Comment:
Should check if `path.c_str() != NULL`. Passing NULL to `dlopen()` is defined behavior and returns handle to current executable